### PR TITLE
Feat: add a built in garbage-collect policy to application

### PR DESCRIPTION
**Pull request by yangsoon**
### Description of your changes

add a built in `garbage-collect` policy, while now only support one option  `keepLegacyResource`, which means gc mechanism will keep the legacy resource when application is updated. you can enable the `keepLegacyResource` options like: 

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: first-vela-app
spec:
  components:
    - name: express-server
      type: webservice
      properties:
        image: crccheck/hello-world
        port: 8000
      traits:
        - type: ingress-1-20
          properties:
            domain: testsvc.example.com
            http:
              "/": 8000
  policies:
    - name: keep-legacy-resource
      type: garbage-collect
      properties:
        keepLegacyResource: true
```

**what will the gc mechanism do when enable the `keepLegacyResource` options**
1. resources with the same GVK and name will be normal updated, and the `OwnerReference` of those resources will refer to the new resourcesTracker.
2. the expired resources and resourceTracker will be keep in cluster.
3. you can delete all resource by delete the application

TODO

- [x] add more test

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

cc @Somefive @wangyikewxgm @captainroy-hy

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/gc.go [hunk 1]
```diff
@@ -59,6 +69,13 @@ func (h *GCHandler) GarbageCollect(ctx context.Context, oldRT, newRT *v1beta1.Re
 		return err
 	}
 	klog.InfoS("Garbage collect for application", "old", h.oldRT.Name, "new", h.newRT.Name)
+
+	// if enabled KeepLegacyResource gc options, GCHandler will:
+	// 1. keep legacy resources created by old application
+	// 2. keep legacy resourceTrackers
+	if h.gcOptions.KeepLegacyResource {
+		return nil
```
Comment (leejanee) [inline pkg/controller/core.oam.dev/v1alpha2/application/dispatch/gc.go:77]: 如果rt没有资源挂钩的时候 还是需要清理掉的

**Comment by hongchaodeng** [comment]
What's the use case? Why keep those legacy resource?

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/gc.go (hunk 1 above)
Comment (yangsoon) [reply pkg/controller/core.oam.dev/v1alpha2/application/dispatch/gc.go:77]: > 如果rt没有资源挂钩的时候 是否应该清理掉

确实需要清理掉没有挂载资源的rt

**Review by wangyikewxgm — APPROVED** [review]
是否可以清理appRev的时候，清理对应对应的resourceTracker

**Comment by yangsoon** [comment]
> 是否可以清理appRev的时候，清理对应对应的resourceTracker

清理resourceTracker会导致历史资源被删除，这里还是希望提供接口让用户主动去删

**Comment by wonderflow** [comment]
Need to write examples(docs) for this policy and usage. The docs don't need to show in our website, but contributors should be able to understand it

**Comment by yangsoon** [comment]
garbage-collect policy usages [doc](https://github.com/oam-dev/kubevela/blob/0c072e4f9a25d2cb7b2cd681c7487ffc7c220e2d/docs/examples/gc-policy/gc.md).

**Review by Somefive — APPROVED** [review]

**Review by wonderflow — APPROVED** [review]
please fix CI

**Review by leejanee — COMMENTED** [review]
please add scenes-oriented test cases.

**Review by leejanee — APPROVED** [review]

**Comment by yangsoon** [comment]
> please add scenes-oriented test cases.

https://github.com/oam-dev/kubevela/blob/213d699b9dd83e5a556aea01469518613b418f38/pkg/controller/core.oam.dev/v1alpha2/application/gc_test.go

**Comment by yangsoon** [comment]
The test coverage results seem to be incorrect, I am sure that the critical path has corresponding tests

**Pull request merged**